### PR TITLE
FT asset controllers; mint/burn;

### PR DESCRIPTION
**Pull request by timohanke**

### CODE REVIEW

File: src/shared/validators.mo [hunk 1]
```diff
@@ -66,7 +67,15 @@ module {
       )) {
         return #err(#FlowsNotSorted);
       };
-      for ((subaccountId, asset) in contribution.inflow.vals()) {
+      let tokenSupplements: Iter.Iter<T.Asset> = u.iterConcat<T.Asset>(
+        Iter.map<(T.SubaccountId, T.Asset), T.Asset>(contribution.inflow.vals(), func (flow) = flow.1),
+        contribution.mints.vals()
+      );
```
Comment (timohanke) [inline src/shared/validators.mo:73]: This should be outflow + mints, not inflow + mints.

The terms inflow and outflow are confusing and we should replace them long term by something better. The terms were meant to be understand form the perspective of the account (out from account = debit and in to account = credit). But using them inside a contribution makes the reader think they are meant to be understood from the perspective of the transaction (like in Bitcoin, input to a transaction and output from a transaction), which is exactly the opposite. 

Let's leave a name change of inflow/outflow for a separate PR.

Let's change the names `tokenSupplements` to `txAssetsIn` and `tokenAbatements` to `txAssetsOut`. The words supplement and abatement are just not that common.

File: src/ledger/ledger_api.mo [hunk 2]
```diff
@@ -51,6 +50,18 @@ actor class LedgerAPI(initialAggregators : [Principal]) {
     };
   };
 
+  // asset interface
+  // create a new fungible token and get the asset id
+  // in the future, calling this will cost a fee in ICP or cycles
+  // an error occurs when the maximum number of asset ids is reached
+  // an error occurs when the call does not carry a valid fee payment
+  // the caller will become the "controller" if the asset id
```
Comment (timohanke) [inline src/ledger/ledger_api.mo:58]: typo: "of the asset id"

**Review by timohanke — COMMENTED** [review]
One change requested in validators.mo

File: src/ledger/ledger.did [hunk 3]
```diff
@@ -82,6 +82,22 @@ type Subaccounts_Result =
...
+  variant {
+    ok;
+    err: variant {NotController; SubaccountNotFound; WrongAssetType}
+  };
+type Burn_Result =
+  variant {
+    ok;
+    err: variant {NotController; SubaccountNotFound; WrongAssetType; InsufficientFunds}
+  };
+
```
[the remark this answers is not in the document]
Comment (AndyGura) [reply src/ledger/ledger.did]: It's already removed

File: src/ledger/ledger.did [hunk 4]
```diff
@@ -120,6 +136,12 @@ service : (vec principal) -> {
 
   addAggregator: (principal) -> (nat);
 
+  // asset interface
+
+  createFungibleToken: () -> (CreateFt_Result); 
+  mint: (SubaccountId, Asset) -> (Mint_Result);
+  burn: (SubaccountId, Asset) -> (Burn_Result);
```
[the remark this answers is not in the document]
Comment (AndyGura) [reply src/ledger/ledger.did]: It's already removed as well

File: src/ledger/ledger_api.mo [hunk 5]
```diff
@@ -51,6 +53,40 @@ actor class LedgerAPI(initialAggregators : [Principal]) {
...
+  public shared (msg) func burn(sid : SubaccountId, asset : Asset) : async Result<(), BurnError> {
+    // check if the asset is of the #ft variant (otherwise return #WrongAsset)
+    // check if msg.caller is the controller of the asset id 
+    // check if sid exists
+    // check if asset id to burn matches asset id in subaccount 
+    // check if available quantity is sufficient
+    // subtract asset from subaccount
+    nyi()
+  };
+
```
[the remark this answers is not in the document]
Comment (AndyGura) [reply src/ledger/ledger_api.mo]: It's already removed as well

File: src/shared/validators.mo (hunk 1 above)
Comment (AndyGura) [reply src/shared/validators.mo]: Right, I confused that. "mint" should be treated as outflow (like someone gave tokens) and "burn" as inflow (like someone received those tokens)

**Review by AndyGura — APPROVED** [review]

**Pull request merged**